### PR TITLE
Add kernel module building

### PR DESCRIPTION
**Pull request by Doridian**
Requires https://github.com/gokrazy-community/crossbuild-armhf/pull/2

This PR adds building some kernel modules, currently the same as the stock kernel (WiFi and Bluetooth).
It also changes the config alterations to be more in-line with the vanilla/stock kernel

### CODE REVIEW

**Comment by oliverpool** [comment]
Thanks for the PR.

Since you are talking about Wifi and Bluetooth, this is not for a raspberry pi first gen, right?

Maybe it would be nice to have `dist` subfolders with the various [configurations](https://www.raspberrypi.com/documentation/computers/linux_kernel.html#default_configuration)

> For Raspberry Pi 1, Zero and Zero W, and Raspberry Pi Compute Module 1 default (32-bit only) build configuration
> KERNEL=kernel
> make bcmrpi_defconfig

> For Raspberry Pi 2, 3, 3+ and Zero 2 W, and Raspberry Pi Compute Modules 3 and 3+ default 32-bit build configuration
> KERNEL=kernel7
> make bcm2709_defconfig

> For Raspberry Pi 4 and 400, and Raspberry Pi Compute Module 4 default 32-bit build configuration
> KERNEL=kernel7l
> make bcm2711_defconfig

> For Raspberry Pi 3, 3+, 4, 400 and Zero 2 W, and Raspberry Pi Compute Modules 3, 3+ and 4 default 64-bit build configuration
> KERNEL=kernel8
> make bcm2711_defconfig

(for the last one, the name of the repo would not be appropriate anymore :see_no_evil: : maybe a good occasion to use git lfs - https://stackoverflow.com/questions/58280942/import-a-golang-module-with-lfs-objects)

**Comment by Doridian** [comment]
My own testing here is done on a Pi Zero (first gen, hence 32-bit OS).
Ideally, this is something releases with archives would be for, but that isn't how gokrazy works, as it expects everything to be go modules.

While LFS sounds nice, if that is done, we need to be very clear about requiring it, and how to configure it on the client end (if any config is necessary, its been a while since I did anything LFS)
Also, I am not sure if GitHub bills for LFS traffic/storage on open-source projects or not.
It would certainly help clone times for development (as you can skip LFS entirely in such case)

However, I feel as if that would be great for a second PR, to not make one change that will be incredibly huge, since I feel the above suggestion could very well build on top of this being done.

File: cmd/compile/main.go [hunk 1]
```diff
@@ -35,6 +33,30 @@ func execCmd(env map[string]string, stdout io.Writer, stderr io.Writer, cmd stri
...
+CONFIG_MODULE_COMPRESS_ZSTD=n
+
+# WiFi
+CONFIG_RFKILL=y
+CONFIG_CFG80211=y
+CONFIG_BRCMFMAC=m
+
+# Bluetooth
+CONFIG_NLMON=y
+CONFIG_BT=m
```
Comment (oliverpool) [inline cmd/compile/main.go:55]: Stupid question: line 90 you disable all modules `(mod2noconfig is not in RPi 5.15)`, but enable `CONFIG_BT*` as modules. 

Do only some module work? Or did I misunderstood the comment line 90?

File: cmd/compile/main.go [hunk 2]
```diff
@@ -65,24 +87,34 @@ func run() error {
...
+	if _, err := f.Write([]byte(configAddendum)); err != nil {
+		return err
+	}
+	if err := f.Close(); err != nil {
+		return err
+	}
+	// fix the config with our addendums
+	if err := dockerRun("make", "olddefconfig"); err != nil {
+		return err
+	}
```
Comment (oliverpool) [inline cmd/compile/main.go]: I think it would be better to adjust the `adjustTextFile` function instead of:
- running `sed`
- appending some lines
- running `make olddefconfig` (which prints out some warnings)


What do you think of a `adjustConfig` function which performs all 3 operations in one go:
```go
// adjustConfig will disable all settings set as module and adjust the given config
func adjustConfig(path string, config map[string]string) {
	// read whole file
	for _, line := range strings.Split(string(b), "\n") {
		key, value, _ := strings.Cut(line, "=")

		if _, ok := config[key]; ok { // custom setting will be appended
			continue
		}
		if value == "m" { // modules should be disabled
			line = key + "=n"
		}

		_, err = w.WriteString(line + "\n")
		if err != nil {
			return err
		}
	}
	for key, value := range config {
		_, err = w.WriteString(key + "=" + value + "\n")
		if err != nil {
			return err
		}
	}
	// flush and close
}
```

Called like this:
```go
	adjustConfig(".config", map[string]string{
		// Basics
		"CONFIG_SQUASHFS": "y",
		"CONFIG_IPV6":     "y",

		// Disable module compression (wifi needs this)
		"CONFIG_MODULE_COMPRESS_NONE": "y",
		"CONFIG_MODULE_COMPRESS_GZIP": "n",
		"CONFIG_MODULE_COMPRESS_XZ":   "n",
		"CONFIG_MODULE_COMPRESS_ZSTD": "n",

		// WiFi
		"CONFIG_RFKILL":   "y",
		"CONFIG_CFG80211": "y",
		"CONFIG_BRCMFMAC": "m",

		// Bluetooth
		"CONFIG_NLMON":          "y",
		"CONFIG_BT":             "m",
		"CONFIG_BT_BCM":         "m",
		"CONFIG_BT_HCIUART":     "m",
		"CONFIG_BT_HCIUART_BCM": "y",
	})
```

File: cmd/compile/main.go [hunk 3]
```diff
@@ -96,11 +128,18 @@ func run() error {
...
+
 	// copy and rename kernel
 	if err = sh.Copy(filepath.Join(dstFolder, "vmlinuz"), filepath.Join(bootFolder, "zImage")); err != nil {
 		return err
 	}
 
+	// copy modules
+	copy.Copy(filepath.Join(kernelFolder, "modules_out"), dstFolder)
+
```
Comment (oliverpool) [inline cmd/compile/main.go]: A `os.Rename` should be sufficient, no?

**Review by oliverpool — COMMENTED** [review]
Looks good! I left a couple of comments, let me know what you think.

The modules end up deeply nested in a `lib/modules/5.15.61` folder structure. Is this expected? (sorry, I have no experience with kernel modules :see_no_evil: )
image

**Comment by oliverpool** [comment]
You are right about LFS. It adds requirements on the client and bills on github...

Regarding gokrazy, as you point out, it would be better to split this in multiple modules (so that you only have to download 1 kernel when running gokrazy). But this can be considered in another PR.

File: cmd/compile/main.go (hunk 1 above)
Comment (Doridian) [reply cmd/compile/main.go:55]: Building all modules takes the build time roughly 4x with the default set the bcm*_defconfig enables, most of which are pointless for gokrazy, so I went with the same set the vanilla kernel builds.

File: cmd/compile/main.go (hunk 2 above)
Comment (Doridian) [reply cmd/compile/main.go]: I am much more comfortable running `olddefconfig` for two reasons:
1. If you enable any options that have dependencies, this will do that correctly and accurately for you, as opposed to random sed magic and then just running the build hoping for the best (a working kernel)
2. This is exactly what the vanilla kernel does, so it seems to be the proven solution (again, we cannot use `mod2noconfig`, since the RPi kernel is too old, so we use `sed`). (See upstream process here: https://github.com/gokrazy/kernel/blob/main/cmd/gokr-build-kernel/build.go#L1092-L1124 )
(Also see, as expected, the barrage of warnings they produce configuring: https://github.com/gokrazy/kernel/runs/8263485742?check_suite_focus=true#step:6:1068 )

File: cmd/compile/main.go (hunk 1 above)
Comment (Doridian) [reply cmd/compile/main.go:55]: (With the exception we use `sed` for editing out the modules, the vanilla kernel does `mod2noconfig`, but the Raspi kernel is too old to have that option in make)

File: cmd/compile/main.go (hunk 3 above)
Comment (Doridian) [reply cmd/compile/main.go]: Possibly. Only reason I opted for a copy is because every other operation done here is a copy as well. I can try and change that.

**Comment by Doridian** [comment]
> The modules end up deeply nested in a `lib/modules/5.15.61` folder structure. Is this expected? (sorry, I have no experience with kernel modules 🙈 )

Yes. While it might look ugly or weird for a single-kernel appliance-like build, this is how the kernel expects them to be nested to load.

File: cmd/compile/main.go (hunk 2 above)
Comment (Doridian) [reply cmd/compile/main.go]: As for the warnings, building the Linux kernel I've kind of gotten used to warnings in the config generation phase, they seem hard to avoid even when you interactively configure the kernel sometimes.
The only way to avoid them and get the comfort of point `1` above would be to swap back to the `adjustFile` approach and then still run `olddefconfig`, which might still yield some warnings (which, to me, is pretty much equivalent to `sed`, except in golang rather than command line)

**Comment by oliverpool** [comment]
Thanks for the explanation regarding `mod2noconfig`. I think I understand it now.

The best way is probably to keep `sed` for now and add a comment that `make -C scripts/kconfig mod2noconfig` will be available on kernel 5.17 or later.

---

Regarding the warnings, I think it is better if we can address the "trivial" ones (duplicate config).

File: cmd/compile/main.go (hunk 3 above)
Comment (oliverpool) [reply cmd/compile/main.go]: I see 2 drawbacks to `copy.Copy`: 
- it adds a dependency
- if leaves the `linux-sources` git repository "unclean" (since `modules_out` is not in its `.gitignore`
=> using `os.Rename` solves both issues (but I would be fine with another solution which addresses these issues :)

**Comment by Doridian** [comment]
> Regarding the warnings, I think it is better if we can address the "trivial" ones (duplicate config).

I would not call them "trivial", as fixing them requires actually editing around in the kernel config, vs a fairly safe append-only approach.
I really don't want to actually add **edits** to the kernel config (already quite unhappy with the `sed` that is required until 5.17).
I feel editing the config is more unsafe than ignoring the warnings by olddefconfig (considering olddefconfig is considered stable, so its behaviour is not expected to change pretty much.

Especially, as I said, the stock gokrazy kernel does exactly this same thing ( https://github.com/gokrazy/kernel/blob/main/cmd/gokr-build-kernel/build.go#L1092-L1124 )
And, doing this same thing with way more config options, causes a number of duplicate warnings:
https://github.com/gokrazy/kernel/runs/8263485742?check_suite_focus=true#step:6:1068

Basically, I used the stock gokrazy kernel as a guideline on how to solve things, considering that's what this is for.

**Comment by oliverpool** [comment]
What do you think of using [`merge_config.sh`](https://unix.stackexchange.com/a/450407) ?

This way we could move the content of the custom config to a dedicated file `.config.gokrazy` and request its merge with `.config`

**Comment by Doridian** [comment]
> What do you think of using [`merge_config.sh`](https://unix.stackexchange.com/a/450407) ?
> 
> This way we could move the content of the custom config to a dedicated file `.config.gokrazy` and request its merge with `.config`

That actually sounds perfect. Just be aware, the warnings will still appear: https://github.com/torvalds/linux/blob/master/scripts/kconfig/merge_config.sh#L125

**Comment by oliverpool** [comment]
I tried it a bit and you are right that some logs are produced (but they don't feel as "scary" as a warning to me :)
```
Using .config as base
Merging .config.gokrazy
Value of CONFIG_SQUASHFS is redefined by fragment .config.gokrazy:
Previous value: CONFIG_SQUASHFS=m
New value: CONFIG_SQUASHFS=y

Value of CONFIG_IPV6 is redefined by fragment .config.gokrazy:
Previous value: CONFIG_IPV6=m
New value: CONFIG_IPV6=y

# some more lines #

#
# configuration written to .config
#

Value requested for CONFIG_CRYPTO_SHA256 not in final .config
Requested value:  CONFIG_CRYPTO_SHA256=m
Actual value:     CONFIG_CRYPTO_SHA256=y

# some more lines #
```

**Comment by Doridian** [comment]
> I tried it a bit and you are right that some logs are produced (but they don't feel as "scary" as a warning to me :)

Currently trying to fight `merge-config.sh`. Even with just the default config I get hundreds of warnings now, just trying to merge the bcm defconfig without any custom additions.
There is so many my terminal scrollback can't even show all of them.

With so many lines, it is hard to tell if any of them pose an actual issue, or not.

**Comment by oliverpool** [comment]
Is it due to the `sed` maybe?
Does running `make olddefconfig` just after `sed` helps in any way?

**Comment by Doridian** [comment]
Good thinking, that fixed it! (Also added some env vars to the kernel compile that, according to the manual, should be set anyway, but absolutely broke this process now without them)

Only thing I had to add `CONFIG_MODULES=y` to the fragment because it seems with all modules off, the `olddefconfig` just disables module support entirely

**Comment by Doridian** [comment]
Somehow I don't have any modules, however...
Correction: I do have one, but not the WiFi ones. It seems running `olddefconfig` before the merge unset some options necessary to build those modules...

**Comment by Doridian** [comment]
If you have any ideas on how to fix this without digging in deep and figuring out all the necessary options, I'd prefer to just go back to the couple warnings and running olddefconfig after appending.

At least that worked without a ton of additional work, and is likely to continue to work.

**Comment by oliverpool** [comment]
Thank you for trying out. I will try to have a look on Monday!

**Comment by oliverpool** [comment]
I think found a solution to all our problems: `./scripts/config --set-val`!

Here is what I ended up with, [banch `module`](https://github.com/gokrazy-community/kernel-rpi-os-32/compare/module): https://github.com/gokrazy-community/kernel-rpi-os-32/commit/93d73bcee7d2f2b022495cf13d8a28c33ffdfb99
- `make bcmrpi_defconfig`
- `sed s/=m$/=n/i`
- `./scripts/config --set-val ...`
- `make olddefconfig` (no more warning and the resulting .config is exactly the same as with your initial commit :tada: )
- ...

(the `os.Rename` at the end was missing error handling - which revealed that a `chown`was missing`)

Everything looks good on my side.

Can you confirm it works for you as well?

**Comment by Doridian** [comment]
Seems to have worked fine for me to produce the wanted kernel modules.

**Comment by oliverpool** [comment]
Release available as tag [`v1.0.2-5.15.61`](https://github.com/gokrazy-community/kernel-rpi-os-32/releases/tag/v1.0.2-5.15.61)

**Comment by oliverpool** [comment]
I just saw the `build` and `source` folders are symlinks:
https://github.com/gokrazy-community/kernel-rpi-os-32/tree/main/dist/lib/modules/5.15.61

apparently they can be ignored: https://stackoverflow.com/a/13586269/3207406

**Comment by Doridian** [comment]
@oliverpool That is correct. You can safely ignore those folders.

The important bits are the `.ko` files in the kernel folder and (some of) the `modules.*` files in the "root"

**Comment by Doridian** [comment]
For being 100% safe this change works, I just put your 1.0.2 build onto my Pi Zero and it works perfectly.

**Comment by oliverpool** [comment]
Everything still works on my side too (Raspberry Pi Model B Rev 2).

Thanks for the PR and your patience :+1:
